app/Http/Controllers/BuildController.php

```diff
@@ -858,24 +837,31 @@ public function viewFiles(): View|RedirectResponse
                     $ext = 'Tb';
                 }
 
-                $xml .= '<filesizedisplay>' . round($filesize) . ' ' . $ext . '</filesizedisplay>';
-                $xml .= '<isurl>' . $uploadFileOrURL->IsUrl . '</isurl>';
-                $xml .= '</uploadfile>';
-            } else {
-                $xml .= '<uploadurl>';
-                $xml .= '<id>' . $uploadFileOrURL->Id . '</id>';
-                $xml .= '<filename>' . htmlspecialchars($uploadFileOrURL->Filename) . '</filename>';
-                $xml .= '</uploadurl>';
+                $file['filesizedisplay'] = round($filesize) . ' ' . $ext;
+
+                $files[] = $file;
             }
         }
 
-        $xml .= '</cdash>';
+        return view('build.files')
+            ->with('build', $this->build)
+            ->with('files', $files)
+            ->with('urls', $urls);
+    }
 
-        return view('cdash', [
-            'xsl' => true,
-            'xsl_content' => generate_XSLT($xml, base_path() . '/app/cdash/public/viewFiles', true),
-            'title' => 'View Files'
-        ]);
+    public function build_file(int $build_id = null, int $file_id = null) : StreamedResponse
+    {
+        $this->setBuildById($build_id);
+
+        // Validate that the file is associated with the build.
+        if (DB::table('build2uploadfile')->where('buildid', $build_id)->where('fileid', $file_id)->doesntExist()) {
+            abort(404, 'File not found');
+        }
```

Comment:
You could consider making an Eloquent model for the `uploadfile` table, and then using Eloquent relationships from the Build model to perform this query.  You could then use the model retrieved to do the logic below.  If you feel that doing that would be outside the scope of this PR, I think that's fine, and we can do it in a separate PR later on.

Reply:
Given that we want to keep this PR small as part of the next 3.2 release candidate, we can defer this and I'll implement this later.